app/controllers/automated_tests_controller.rb

```diff
@@ -35,14 +35,34 @@ def student_interface
     @student = current_role
     @grouping = @student.accepted_grouping_for(@assignment.id)
 
-    unless @grouping.nil?
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        last_student_run = @grouping.test_runs.where(role: @grouping.accepted_students).first
+        if last_student_run.nil?
+          @next_token_generation_time = @assignment.token_start_date + @assignment.token_period.hours
+        else
+          hours_from_start = (Time.current - @assignment.token_start_date) / 3600
+          if @assignment.non_regenerating_tokens
```

Comment:
This is a good case to consider, but should actually be handled separately. When tokens do not regenerate, or when there are *unlimited tokens*, we should not calculate a value for `@next_token_generation_time` at all, but instead just set it to `nil`. This can be checked in the condition on line 45.